lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -1458,40 +1510,67 @@ deleteDelegationCertificates wid filters = do
+            ] []
+    forM utxos $ \utxo -> do
+        (utxo, ) . fmap entityVal <$> selectList
+            [ UtxoTokenWalletId ==. utxoWalletId utxo
+            , UtxoTokenSlot ==. utxoSlot utxo
+            , UtxoTokenTxId ==. utxoInputId utxo
+            , UtxoTokenTxIndex ==. utxoInputIndex utxo
+            ] []
 
--- This relies on available information from the database to reconstruct
--- coin selection information for __outgoing__ payments. We can't however guarantee
+-- This relies on available information from the database to reconstruct coin
+-- selection information for __outgoing__ payments. We can't however guarantee
 -- that we have such information for __incoming__ payments (we usually don't
 -- have it).
 --
--- To reliably provide this information for incoming payment, it should be looked
--- up when applying blocks from the global Ledger, but that is future work
+-- To reliably provide this information for incoming payments, it should be
+-- looked up when applying blocks from the global ledger, but that is future
+-- work.
 --
 -- See also: issue #573.
 selectTxs
     :: [TxId]
-    -> SqlPersistT IO ([(TxIn, Maybe TxOut)], [TxOut], [TxWithdrawal])
+    -> SqlPersistT IO
+        ( [(TxIn, Maybe (TxOut, [TxOutToken]))]
+        , [(TxOut, [TxOutToken])]
+        , [TxWithdrawal]
+        )
 selectTxs = fmap concatUnzip . mapM select . chunksOf chunkSize
   where
     select txids = do
         inputs <- fmap entityVal <$> selectList
             [TxInputTxId <-. txids]
             [Asc TxInputTxId, Asc TxInputOrder]
 
-        resolvedInputs <- toOutputMap . fmap entityVal <$>
-            combineChunked inputs (\inputsChunk -> selectList
-                [TxOutputTxId <-. (txInputSourceTxId <$> inputsChunk)]
-                [Asc TxOutputTxId, Asc TxOutputIndex])
-
-        outputs <- fmap entityVal <$> selectList
-            [TxOutputTxId <-. txids]
-            [Asc TxOutputTxId, Asc TxOutputIndex]
+        resolvedInputs <- fmap toOutputMap $
+            combineChunked inputs $ \inputsChunk -> do
+                outs <- fmap entityVal <$> selectList
+                    [TxOutputTxId <-. (txInputSourceTxId <$> inputsChunk)]
+                    [Asc TxOutputTxId, Asc TxOutputIndex]
+                forM outs $ \out ->
+                    (out,) . fmap entityVal <$> selectList
+                        [ TxOutTokenTxId ==. txOutputTxId out
+                        , TxOutTokenTxIndex ==. txOutputIndex out
+                        ]
+                        []
+
+        outputs <- do
+            outs <- fmap entityVal <$> selectList
+                [TxOutputTxId <-. txids]
+                [Asc TxOutputTxId, Asc TxOutputIndex]
+            forM outs $ \out ->
+                (out,) . fmap entityVal <$> selectList
+                    [ TxOutTokenTxId ==. txOutputTxId out
+                    , TxOutTokenTxIndex ==. txOutputIndex out
+                    ]
+                    []
```

Comment:
> The part fetching the output + token is identical (modulo the list of tx ids) here and above. I'd suggest factoring out the common part in a dedicated function to make avoid duplication and clearly show the commonality.

Well spotted! I didn't notice that.

Fixed in 356766fcc1570760d189f567979aa0301b65473f.

Reply:
> On a different note, this number of requests that this generates seems rather large. I haven't yet checked the database benchmarks but we should make sure to measure this correctly and generate bench database with token bundles.

Great point.

The DB benchmark currently generates ada-only bundles.

I'll upgrade this to generate bundles that include non-ada tokens.

Anecdotally, when running the DB property tests, I do see a small lengthening in the total run-time (around 5% to 10%). However, this could be because generating token bundles is inherently more time-consuming than just generating coins.

Reply:
@jonathanknowles wrote:
> I'll upgrade this to generate bundles that include non-ada tokens.

Just for the record, we decided to handle the upgrade of the DB benchmarks in a separate ticket, which I have created here: https://jira.iohk.io/browse/ADP-627